private/bufpkg/bufimage/bufimagebuild/builder.go

```diff
@@ -81,132 +79,63 @@ func (b *builder) build(
-		fileAnnotations = append(fileAnnotations, buildResult.FileAnnotations...)
-	}
-	if len(fileAnnotations) > 0 {
-		return nil, bufanalysis.DeduplicateAndSortFileAnnotations(fileAnnotations), nil
+	if len(buildResult.FileAnnotations) > 0 {
+		return nil, bufanalysis.DeduplicateAndSortFileAnnotations(buildResult.FileAnnotations), nil
 	}
 
-	descFileDescriptors, err := getDescFileDescriptorsFromBuildResults(buildResults, paths)
+	fileDescriptors, err := checkAndSortFileDescriptors(buildResult.FileDescriptors, paths)
 	if err != nil {
 		return nil, nil, err
 	}
 	image, err := getImage(
 		ctx,
-		excludeSourceCodeInfo,
-		descFileDescriptors,
+		options,
+		fileDescriptors,
 		parserAccessorHandler,
-		getSyntaxUnspecifiedFilenamesFromBuildResults(buildResults),
-		getFilenameToUnusedDependencyFilenamesFromBuildResults(buildResults),
+		buildResult.SyntaxUnspecifiedFilenames,
+		buildResult.FilenameToUnusedDependencyFilenames,
 	)
 	if err != nil {
 		return nil, nil, err
 	}
 	return image, nil, nil
 }
 
-func getBuildResults(
-	ctx context.Context,
-	parserAccessorHandler bufmoduleprotoparse.ParserAccessorHandler,
-	paths []string,
-	excludeSourceCodeInfo bool,
-) []*buildResult {
-	ctx, span := trace.StartSpan(ctx, "parse")
-	defer span.End()
-
-	chunks := chunkPaths(paths)
-	var buildResults []*buildResult
-	buildResultC := make(chan *buildResult, len(chunks))
-	for _, iPaths := range chunks {
-		iPaths := iPaths
-		go func() {
-			var buildResult *buildResult
-			defer func() {
-				select {
-				case buildResultC <- buildResult:
-				case <-ctx.Done():
-				}
-			}()
-			defer func() {
-				// Recover any panics here since we run in a goroutine
-				v := recover()
-				if v != nil {
-					buildResult = newBuildResult(
-						nil,
-						nil,
-						nil,
-						nil,
-						fmt.Errorf("panic: %v, stack:\n%s", v, string(debug.Stack())),
-					)
-				}
-			}()
-			buildResult = getBuildResult(
-				ctx,
-				parserAccessorHandler,
-				iPaths,
-				excludeSourceCodeInfo,
-			)
-		}()
-	}
-	for i := 0; i < len(chunks); i++ {
-		select {
-		case <-ctx.Done():
-			return []*buildResult{newBuildResult(nil, nil, nil, nil, ctx.Err())}
-		case buildResult := <-buildResultC:
-			buildResults = append(buildResults, buildResult)
-		}
-	}
-	return buildResults
-}
-
 func getBuildResult(
 	ctx context.Context,
-	parserAccessorHandler bufmoduleprotoparse.ParserAccessorHandler,
+	parserAccessorHandler bufmoduleprotocompile.ParserAccessorHandler,
 	paths []string,
-	excludeSourceCodeInfo bool,
+	options *buildOptions,
 ) *buildResult {
-	var errorsWithPos []protoparse.ErrorWithPos
-	var warningErrorsWithPos []protoparse.ErrorWithPos
-	var lock sync.Mutex
-	parser := protoparse.Parser{
-		IncludeSourceCodeInfo: !excludeSourceCodeInfo,
-		Accessor:              parserAccessorHandler.Open,
-		ErrorReporter: func(errorWithPos protoparse.ErrorWithPos) error {
-			// protoparse isn't concurrent right now but just to be safe
-			// for the future
```

Comment:
`protocompile` _is_ concurrent. However, this explicit locking is unnecessary as it handles thread-safe access to the given reporter internally.
https://pkg.go.dev/github.com/bufbuild/protocompile@v0.0.0-20221004142453-757ef5688631/reporter#Reporter